test/framework/clusterstatevalidations/cluster.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	for _, node := range cpNodes.Items {
+		if err := validateNodeReady(node, clus.Spec.KubernetesVersion); err != nil {
+			return fmt.Errorf("failed to validate controlplane %s", err)
+		}
+	}
+	return nil
+}
+
+// ValidateWorkerNodes retries the worker nodes from the cluster and checks them against the cluster.Spec.
+func ValidateWorkerNodes(ctx context.Context, vc clusterf.StateConfig) error {
+	clus := vc.ClusterSpec.Cluster
+	nodes := &corev1.NodeList{}
+	if err := vc.ClusterClient.List(ctx, nodes); err != nil {
+		return fmt.Errorf("failed to list nodes %s", err)
+	}
+	wn := clus.Spec.WorkerNodeGroupConfigurations
+	// deduce the worker node group configuration to node mapping via the machine deployment and machine set
+	for _, w := range wn {
+		workerGroupCount := 0
+		ms, err := getWorkerNodeMachineSets(ctx, vc, w)
+		if err != nil {
+			return fmt.Errorf("failed to get machine sets when validating worker node: %v", err)
+		}
+		for _, node := range nodes.Items {
+			ownerName, ok := node.Annotations["cluster.x-k8s.io/owner-name"]
+			if ok {
+				// there will be multiple machineSets present on a cluster following an upgrade.
+				// find the one that is associated with this worker node, and execute the validations.
+				for _, machineSet := range ms {
+					if ownerName == machineSet.Name {
+						workerGroupCount++
+						if err := validateNodeReady(node, vc.ClusterSpec.Cluster.Spec.KubernetesVersion); err != nil {
+							return fmt.Errorf("failed to validate worker node ready %v", err)
+						}
+					}
+				}
+			}
+		}
+		if workerGroupCount != *w.Count {
+			return fmt.Errorf("worker node group %s count does not match expected: %d of %d", w.Name, workerGroupCount, *w.Count)
+		}
+	}
+	return nil
+}
+
+// ValidateClusterDoesNotExist checks that the cluster does not exist by attempting to retrieve the CAPI cluster.
+func ValidateClusterDoesNotExist(ctx context.Context, vc clusterf.StateConfig) error {
+	clus := vc.ClusterSpec.Cluster
+	capiCluster, err := controller.GetCAPICluster(ctx, vc.ManagementClusterClient, clus)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster %s", err)
+	}
+	if capiCluster != nil {
+		return fmt.Errorf("cluster %s exists", capiCluster.Name)
+	}
+	return nil
+}
+
+// ValidateCilium gets the cilium-config from the cluster and checks that the cilium
+// policy in cluster.Spec matches the enabled policy in the config.
+func ValidateCilium(ctx context.Context, vc clusterf.StateConfig) error {
+	clusterClient := vc.ClusterClient
+
+	yaml := vc.ClusterSpec.Cluster
+	cm := &corev1.ConfigMap{}
+	key := types.NamespacedName{Namespace: "kube-system", Name: "cilium-config"}
+	err := clusterClient.Get(ctx, key, cm)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve configmap %s", err)
+	}
+
+	clusterCilium := cm.Data["enable-policy"]
+	yamlCilium := string(yaml.Spec.ClusterNetwork.CNIConfig.Cilium.PolicyEnforcementMode)
```

Comment:
is there any way we can leverage the code in the cilium package to do this?
I believe we have code to do all this
Plus we could benefit from checking the version as well, useful to for upgrades

Reply:
Possible we can mark this down for improvement in another PR? This one is just moving the validations.